src/charm.py

```diff
@@ -0,0 +1,331 @@
+        if not self._container.can_connect():
+            self.unit.status = WaitingStatus("Waiting for container to be ready")
+            event.defer()
+            return
+        self._write_ue_config_file()
+
+    def _configure_sdcore_smf(self, event: EventBase) -> None:
+        """Adds pebble layer and manages Juju unit status.
+
+        Args:
+            event: Juju event
+        """
+        if not self._database_relation_is_created:
+            self.unit.status = BlockedStatus("Waiting for database relation to be created")
+            return
+        if not self._smf_database_relation_is_created:
+            self.unit.status = BlockedStatus("Waiting for SMF database relation to be created")
+            return
+        if not self._nrf_relation_is_created:
+            self.unit.status = BlockedStatus("Waiting for NRF relation to be created")
+            return
+        if not self._container.can_connect():
+            self.unit.status = WaitingStatus("Waiting for container to be ready")
+            return
+        if not self._database_is_available:
+            self.unit.status = WaitingStatus("Waiting for database relation to be available")
+            return
+        if not self._smf_database_is_available:
+            self.unit.status = WaitingStatus("Waiting for SMF database relation to be available")
+            return
+        if not self._nrf_is_available:
+            self.unit.status = WaitingStatus("Waiting for NRF relation to be available")
+            return
+        if not self._ue_config_file_is_written:
+            event.defer()
+            self.unit.status = WaitingStatus(
+                f"Waiting for `{UEROUTING_CONFIG_FILE}` config file to be pushed to workload container"  # noqa: W505, E501
+            )
+            return
+        if not self._config_file_is_written:
+            logger.warning("Writing config file")
+            self._write_config_file(
+                database_url=self._smf_database_data["uris"].split(",")[0],
+                nrf_url=self._nrf_requires.get_nrf_url(),
+            )
+        self._configure_pebble()
+
+    def _configure_pebble(self) -> None:
+        """Adds layer to pebble config if the proposed config is different from the current one."""
+        self._container.add_layer("smf", self._pebble_layer, combine=True)
+        self._container.replan()
+        self.unit.status = ActiveStatus()
+
+    def _write_config_file(self, database_url: str, nrf_url: str) -> None:
+        """Writes config file to workload.
+
+        Args:
+            database_url (str): Database URL
+            nrf_url (str): NRF URL
+        """
+        jinja2_env = Environment(loader=FileSystemLoader("src/templates"))
+        template = jinja2_env.get_template("smfcfg.yaml.j2")
+        content = template.render(
+            database_name=DATABASE_NAME,
+            database_url=database_url,
+            smf_url=self._smf_hostname,
+            smf_sbi_port=SMF_SBI_PORT,
```

Comment:
I'd pass those as parameters as well instead of having hald the attributes be parameters and the others directly accessed.